### ISSUE TITLE
Functional single file component with components option.

### ISSUE THREAD

**Issue by inuscript** [vuejs/vue#7492]
### Version
2.5.13

### Reproduction link

NG pattern (functional)
https://codesandbox.io/s/004vv2onw0

OK pattern (no functional)
https://codesandbox.io/s/q9k5q8qq56

### Steps to reproduce
I found can't use `components` option when  `functional` single file component.

```html
<template functional>
  <div>
    <some-children />
  </div>
</template>

<script>
import SomeChildren from "./SomeChildren"

export default {
  components: {
    SomeChildren
  }
}
</script>
```

It's occure `Unknown custom element`.

### What is expected?
Not occure `Unknown custom element` and use child component

### What is actually happening?
It's occure `Unknown custom element`

---

In workaround, it not occure when use `Vue.component`.

```js
import Vue from "vue"
import SomeChildren from "./SomeChildren"
Vue.component("some-children", SomeChildren);

export default {}

// can use  <some-children />
```

**Comment by jingle2008** [comment]
Ran into exact same issue here, nice workaround!!!

**Comment by gregolai** [comment]
Just learning Vue and struggling with this for the past few hours! Thanks for bringing it up!

**Comment by ywwhack** [comment]
Here is another workaround, **it avoids global component**, but looks not pretty

```js
<template functional>
  <div>
    <component :is="props.components.SomeChildren"></component>
  </div>
</template>

<script>
import SomeChildren from "./SomeChildren.vue";
export default {
  props: {
    components: {
      type: Object,
      default() {
        return {
          SomeChildren
        };
      }
    }
  }
};
</script>
```

**Comment by alexsasharegan** [comment]
I think it's worth mentioning here that the error message is quite unintuitive. The `Unkown custom element` error bubbles up to the first instance component. If for some reason the feature for local functional component registration does not get implemented, at least add a dev warning that says something to the effect of `Invalid property "components" on functional component X`.

**Comment by darkylmnx** [comment]
any update on implementing components option in functional components ?

The workarounds are good but that seems pretty hacky IMO especially when mentioning that `<component :is="injections.components.SomeChildren"></component>` must be fore dynamic components and not for known components

**Comment by KumaCool** [comment]
parent:
```javascript
// template functional 
    v-list
        component(
            :is='injections.components.myListItem'
            v-for='item in props.data',
            :key='item.title',
            :data='item')

// script
import myListItem from './listItem'
export default {
    name: 'myList',
    inject: {
        components: {
            default: {myListItem}
        }
    }
}
```
children:
```javascript
// template functional
    v-list-group(v-if='Array.isArray(props.data.children)')
        v-list-tile(slot='activator')
            v-list-tile-content
                v-list-tile-title {{props.data.title}}
        my-list-item(
            v-for='item in props.data.children',
            :key='item.title',
            :data='item')
    v-list-tile(v-else)
        v-list-tile-content
            v-list-tile-title {{props.data.title}}

// script
export default {
    name: 'myListItem'
}
```
app:
```javascript
// template
    #app
        my-list(:data='list')

// script
import myList from './list'
export default {
    name: 'App',
    components: {myList},
    data() {
        return {
            list: [
                {title: 1},
                {title: 2},
                {title: 3},
                {
                    title: 4,
                    children: [
                        {title: 41},
                        {title: 42},
                        {title: 43}
                    ]
                },
                {title: 5}
            ]
        }
    }
}

```
children error: Unknown custom element
multi-level functional components nesting failed.
help me!~

**Comment by Alendorff** [comment]
Any ideas how to implement dynamic async components inside functional components? 

E.g. functional should be simple wrapper like 

```vue
<component :is="componentName" /> 
```

where `componentName` is one of the dynamically imported components?

For non-functional component it looks like this:

```vue
<template>
  <component :is="componentName" v-bind="$attrs"/>
</template>

<script>
const someCondition = Math.random() > 0.5

export default {
  name: 'PolicyRequestInfo',
  components: {
    FirstDynamic: () => import('./FirstDynamic'),
    SecondDynamic: () => import('./SecondDynamic')
  },
  computed: {
    componentName() {
      return SomeCondition ? 'FirstDynamic' : 'SecondDynamic'
    }
  }
}
</script>
```

For functional component I have no idea how to make it work.

**Comment by meteorlxy** [comment]
https://github.com/vuejs/vue/blob/70754084ec2d84e44effeb8f82efda3397a08b0d/types/options.d.ts#L123-L133

The type declaration shows that, functional component of current version can't accept `component` option in fact.

Of course a possible solution is to use the render function:

```js
<script>
import SomeChildren from './SomeChildren.vue'

export default {
  render (h) {
	return h('div', [
      h(SomeChildren),
    ])
  }
}
</script>
```

**Comment by Alendorff** [comment]
Yep, it actually works with render functions. I use something like that now:

```vue
<script>
const component = Math.random() > 0.5 
    ? () => import('./compA') 
    : () => import('./compB')

export default {
  functional: true,
  render(h, context) {
    return h(component, context.data, context.children)
  }
}
</script>
```

**Comment by ycmjason** [comment]
Would really love to see `components` option being supported for functional components.

**Comment by dcwarwick** [comment]
This feature would certainly extend the usefulness of functional components in a natural way. I for one had not realised that `components` _wasn't_ supported in functional components until we tried it and found out. We have a library of components as SFCs, and wanted to mark some of the simple ones with no internal state as functional.

At the moment, if a component is a SFC and has child component dependencies and we want to mark it functional, the options would seem to be:
* convert the template to a render function
* globally register all the child components required
* an ingenious workaround that loads the dependency child components using `props` or `inject` and then reference them with a `<component :is="...">`

However, in a library of many interlinked components all done as SFCs it is desirable to retain the template for consistency with the other components, and undesirable to globally register names. The workaround for all its ingenuity looks rather messy/fiddly and would require explanation for maintenance. It would be much neater just to be able to declare child components with `components` exactly as for non-functional SFCs.

Fix #8143 looks good and ready to go. If there isn't a reason not to, can it be delivered? We've held off from making any of our components functional atm, and would use it right away :-)

**Comment by decademoon** [comment]
> I for one had not realised that `components` *wasn't* supported in functional components until we tried it and found out.

That was my experience too. I remember being quite surprised when I discovered components weren't supported in functional components. This really, *really* diminished the usefulness of functional components and I rarely use them because of this.

I do use JSX in my Vue projects because sometimes I need to drop down into the render function, so this isn't technically a huge issue for me, but I like to avoid JSX whenever possible.

I'm hoping Vue 3 will improve on this in some way. React already does functional components well.

> an ingenious workaround described above by @ywwhack @caikan and others that loads the dependency child components using `props` or `inject` and then references them with a `<component :is="...">`

I'm not a fan of this hack TBH. I will just stay clear of functional components until this issue is officially resolved.

**Comment by ffxsam** [comment]
@yyx990803 @LinusBorg Any plans to fix this before Vue 3.0?

**Comment by ShamatienkoYaroslav** [comment]
Found some another way:

`<template>
  <component :is="$options.SimpleCard"
    hover
    pointer
    :active="active"
    @click="$emit('select')"
  >
    <div>
      <div class="font-semibold mb-2">{{ item.name }}</div>
      <div class="text-gray-800 text-xs">{{ item.fullAddress }}</div>
    </div>
  </component>
</template>

<script>
import SimpleCard from '@/components/Common/SimpleCard'

export default {
  name: 'th-pq-warehouses-catalog-list-item',

  props: {
    item: {
      type: Object,
      required: true,
    },
    active: Boolean,
  },

  SimpleCard
}
</script>
`

But `component :is=` looks not good.

**Comment by sisou** [comment]
I just made a SFC with `<template functional>` and a `components` option in the script object, and there are no complaints by the build system or during runtime.

This is with Vue 2.6.6.

**Comment by tettoffensive** [comment]
Is the workaround from @ywwhack / @caikan  workaround still the best way to go these days? With all the mentions at the end it's hard to tell if there's an official fix.

**Comment by JaredDahlke** [comment]
+1. Will not be using the work around as it decreases code readability (since it cannot be found in Vue documentation). Please implement this much needed fix. Thank you

**Comment by douira** [comment]
I suspect this, like some other features for functional components, will not be implemented since Vue 3 (the alpha/proposal) alleviates the whole issue by making the difference between functional and non-functional components small. See this and the following comments: https://github.com/vuejs/vue/pull/8143#issuecomment-482481117

**Comment by madebycaliper** [comment]
hey @maksnester I'm working on this for a Vue 2.6.x project. I'm trying to allow the functional component to accept a `name` prop to make it even more dynamic. Any clue on how I can implement your solution dynamically?

Asked on Stackoverflow too ([here](https://stackoverflow.com/questions/61668562/advanced-vue-js-dynamic-functional-component-using-is-syntax-and-render-funct)) if you want more details about my problem. Thanks!

**Comment by privatenumber** [comment]
As an alternative option, [vue-import-loader](https://github.com/privatenumber/vue-import-loader) offers support for component resolution in functional components

**Comment by darkylmnx** [comment]
Still no improvements on this till now?
I found a future-proof way to do it but still not happy with it:
```vue
<script>
import ChildComp from '@/components/ChildComp.vue';
export default {
  components: {
    ChildComp,
  },
};
</script>
<template>
  <component :is="$options.components.ChildComp" any-prop="value" />
</template>
```

**Comment by andreasvirkus** [comment]
@darkylmnx it's a #wontfix since Vue 3 will make functional components irrelevant, as I understood.

**Comment by darkylmnx** [comment]
Oh... I see.

**Comment by webcoderkz** [comment]
There's RFC: https://github.com/vuejs/rfcs/blob/master/active-rfcs/0007-functional-async-api-change.md

We will just rewrite our functional components as plain functions, that's all, much better, imo.
